sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePartitionedUploader.cs

```diff
@@ -0,0 +1,370 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+using Azure.Storage.Files.DataLake.Models;
+using Azure.Storage.Shared;
+
+namespace Azure.Storage.Files.DataLake
+{
+    internal class DataLakePartitionedUploader
```

Comment:
Could we have either a generic shared source base class or an internal interface to avoid duplicating this code?  There's _a lot_ of complex logic here I'd love to have just a single copy of.

Reply:
I moved as much logic as I could to ChunkedStream and PartitionedUploadExtensions in common.  I don't see how we can generalize the rest of the logic without adding common overloads to DataLakeFileClient and BlobClient for Staging/Appending and Commiting/Flushing data.

Reply:
We could make that work with something like a shared source 

```C#
internal interface ISupportChunkedUpload
{
    void AddChunk(...);
    void SaveChunks(...);
}
```

and then implement it explicitly with

```C#
public partial class BlobClient : ISupportChunkedUpload
{
    void ISupportChunkedUpload.AddChunk(...) => StageBlock(...);
    void ISupportChunkedUpload.SaveChunks(...) => CommitBlockList(...);
}

public partial class DataLakeFileClient : ISupportChunkedUpload
{
    void ISupportChunkedUpload.AddChunk(...) => Append(...);
    void ISupportChunkedUpload.SaveChunks(...) => Flush(...);
}
```
which would add no public APIs but still present a common interface to `PartitionedUploader`.

I'm sure it would take more work than just this _(dealing with the different return types, etc., would probably require the interface to be generic)_, but it would be really great from an engineering point of view.

Reply:
How would we handle the different parameters?  Append and Flush takes numerical offsets, while Stage and Commit takes commit IDs.  Also, in the actual business logic we need to keep track of offsets when dealing with Data Lake, but not Blobs.

Can public classes (BlobClient and DataLakeClient) extend an internal class?

I get where you are coming from with reducing duplicate code, but I worry the PartitionedUploader is going to become super generic and hard to read.

Reply:
I'm going to hold off on attempting to create a unified PartitionedUploader for now.  There are subtle differences in the business logic between Blob and Data Lake Upload.

Reply:
Okay - filed https://github.com/Azure/azure-sdk-for-net/issues/10026 to track.